### PR TITLE
SVM: Hoist program cache recompilation up from Bank

### PR DESCRIPTION
**Pull request by buffalojoec**
#### Problem
The program cache is an integral part of the SVM, but its API is being driven both
by Bank and SVM. Since the program cache exists directly on the SVM's
`TransactionBatchProcessor`, it makes sense to include any necessary program
cache operations within SVM.

#### Summary of Changes

- First, hoist the recompilation code up to SVM as a public API function, to be
   invoked by Bank.
- Second, rename "recompilation" to "cache preparation".

### CODE REVIEW

**Comment by pgarg66** [comment]
@buffalojoec , since this PR includes refactor of recompilation phase, let's wait for either @Lichtso or @alessandrod (or both) to take a look at it before merging it.

**Comment by buffalojoec** [comment]
Note we can probably port more program-cache related items, such as `replenish_program_cache` into this new `program_cache` module!

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -1108,9 +1109,18 @@ impl Bank {
 
         let (epoch_stakes, epoch_stakes_time_us) = measure_us!(parent.epoch_stakes.clone());
 
-        let (transaction_processor, builtin_program_ids_time_us) = measure_us!(
-            TransactionBatchProcessor::new_from(&parent.transaction_processor, slot, epoch)
-        );
+        let (transaction_processor, new_transaction_processor_timings) =
+            TransactionBatchProcessor::new_from(
+                &parent.transaction_processor,
+                parent.as_ref(),
```
Comment (Lichtso) [inline runtime/src/bank.rs]: The accounts were loaded from `new` not from `parent`.

Comment (buffalojoec) [reply runtime/src/bank.rs]: Ahh, I see the comment about touching AccountsDB in the epoch update step. Good call, likely better to use `new`. Done!

File: svm/src/program_cache.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    solana_program_runtime::loaded_programs::{ForkGraph, ProgramCache},
+    solana_sdk::{
+        clock::{Epoch, Slot},
+        epoch_schedule::EpochSchedule,
+        feature_set::FeatureSet,
+    },
+    std::sync::{atomic::Ordering::Relaxed, Arc, RwLock},
+};
+
+pub(crate) fn recompile_program_cache<CB: TransactionProcessingCallback, FG: ForkGraph>(
```
Comment (Lichtso) [inline svm/src/program_cache.rs]: Should probably use this opportunity to also rename it to `preheat_program_cache_for_upcoming_feature_set` or something. Because it is not just about compilation, it is:
- Account loading
- ELF parsing
- Verification
- Relocation
- Compilation

Comment (buffalojoec) [reply svm/src/program_cache.rs]: Something about `preheat` isn't sitting with me. What about `warm_up_program_cache_for_upcoming_feature_set`?

Comment (dmakarov) [reply svm/src/program_cache.rs]: `prepare_program_cache_for_upcoming_feature_set`

Comment (buffalojoec) [reply svm/src/program_cache.rs]: I was going to suggest `prepare`, hah, but warm up sounded cool, too. "Warm up" also has a very, very slight implication of the process being incremental.

Comment (dmakarov) [reply svm/src/program_cache.rs]: I prefer easy to understand and precise in meaning to _cool_

Comment (Lichtso) [reply svm/src/program_cache.rs]: There is also one or two tests which need to be renamed.

Comment (buffalojoec) [reply svm/src/program_cache.rs]: @Lichtso, I only saw one test to rename. Lmk if you know of more. Included in my latest commit.

Comment (Lichtso) [reply svm/src/program_cache.rs]: And there a few comments in loaded_programs.rs now, which also should be adjusted.

Comment (buffalojoec) [reply svm/src/program_cache.rs]: It would be great if you could be more specific.

Comment (Lichtso) [reply svm/src/program_cache.rs]: If you search for `recompilation` in https://github.com/anza-xyz/agave/blob/master/program-runtime/src/loaded_programs.rs there are 8 mentions which should be renamed.

Comment (buffalojoec) [reply svm/src/program_cache.rs]: Gotcha, didn't realize we were renaming the phase, too. Done in latest commit.

**Comment by buffalojoec** [comment]
@Lichtso I've rebased this onto your latest cache changes, and my latest runtime changes. I've also squashed a couple of items.

Squashed:
- I'm using `new` from the very beginning, as requested in initial review.
- Squashed all `recompilation` renames together.

I also decided to save the introduction of the `program_cache` module for another PR.

File: runtime/src/bank.rs [hunk 3]
```diff
@@ -1224,76 +1236,6 @@ impl Bank {
...
-                    .min(slots_in_epoch)
-                    .checked_div(2)
-                    .unwrap();
-            let mut program_cache = new.transaction_processor.program_cache.write().unwrap();
-            if program_cache.upcoming_environments.is_some() {
-                if let Some((key, program_to_recompile)) = program_cache.programs_to_recompile.pop()
-                {
-                    let effective_epoch = program_cache.latest_root_epoch.saturating_add(1);
-                    drop(program_cache);
-                    if let Some(recompiled) = new.load_program(&key, false, effective_epoch) {
```
Comment (Lichtso) [inline runtime/src/bank.rs:1241]: With this `Bank::load_program()` is only used in ledger-tool I think. But that can be removed in a separate PR.

**Review by Lichtso — DISMISSED** [review]

**Comment by buffalojoec** [comment]
Sorry @Lichtso but out of absolute safety, I've decided to keep the call site in Bank where it was originally, and drop the folding of the recompilation step into `new_from`, at least until we can continue some of our testing/refactoring work on the global cache.

File: runtime/src/bank.rs (hunk 3 above)
Comment (buffalojoec) [reply runtime/src/bank.rs:1241]: I'll add a PR for this when this one lands.

**Review by Lichtso — DISMISSED** [review]

**Comment by buffalojoec** [comment]
@pgarg66 @dmakarov Can you guys have another look now that @Lichtso has approved?

**Review by dmakarov — APPROVED** [review]
LGTM
